internal/pkg/webconfig/webconfig.go

```diff
@@ -0,0 +1,189 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+package webconfig
+
+import (
+	"context"
+	"crypto/tls"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net"
+	"net/http"
+	"sync"
+	"time"
+
+	ttls "github.com/talos-systems/crypto/tls"
+	"github.com/talos-systems/crypto/x509"
+	tnet "github.com/talos-systems/net"
+)
+
+const indexPage = `
+<html>
+	<head>
+		<title>Talos Interactive Configurator</title>
+	</head>
+	<body>
+	  <h3>Talos Interactive Configuration</h3>
+	  <form action="/" method="POST">
+	  <p>Upload configuration: <input type="file" name="config" />
+	  </form>
+	</body>
+</html>
+`
+
+// Server defines an http-based configuration receiver service.
+type Server struct {
+	logger *log.Logger
+
+	configBytes []byte
+
+	returnSignal chan struct{}
+
+	serverError error
+
+	svr *http.Server
+
+	mu sync.Mutex
+}
+
+// Stop tells the server to shut down.
+func (s *Server) Stop() {
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+	defer cancel()
+
+	s.mu.Lock()
+	if s.returnSignal != nil {
+		close(s.returnSignal)
+
+		s.returnSignal = nil
+	}
+
+	if s.svr != nil {
+		s.svr.Shutdown(ctx) // nolint: errcheck
+	}
+	s.mu.Unlock()
+}
+
+// ServeHTTP implements http.Handler.
+func (s *Server) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	switch r.Method {
+	case "GET":
+		s.handleGet(w, r)
+	case "POST":
+		s.handlePost(w, r)
+	default:
+		w.WriteHeader(http.StatusBadRequest)
+	}
+}
+
+// nolint: unparam
+func (s *Server) handleGet(w http.ResponseWriter, r *http.Request) {
+	w.WriteHeader(http.StatusOK)
+
+	_, err := w.Write([]byte(indexPage))
+	if err != nil {
+		s.logger.Printf("webconfig: failed to write index page: %s", err.Error())
+	}
+}
+
+func (s *Server) handlePost(w http.ResponseWriter, r *http.Request) {
+	f, header, err := r.FormFile("config")
+	if err != nil {
+		s.logger.Println("webconfig: failed to read config POST:", err)
+		w.WriteHeader(http.StatusBadRequest)
+
+		return
+	}
+
+	if header.Size < 1 {
+		w.WriteHeader(http.StatusBadRequest)
+
+		return
+	}
+
+	s.configBytes, err = ioutil.ReadAll(f)
+	if err != nil {
+		s.logger.Println("webconfig: failed to read file from config POST:", err)
+		w.WriteHeader(http.StatusInternalServerError)
+
+		return
+	}
+
+	if _, err = w.Write([]byte(`{ "accepted": "ok"}`)); err != nil {
```

Comment:
we might want to set Content-Type here explicitly to make it display nicely?

or should it be HTML in fact to show that nicely in the web UI?

Reply:
Yeah, it probably should be just plain html.